File: pascal-api-examples/tts/zipvoice-zh-en.pas
{ Copyright (c)  2026  Xiaomi Corporation }
program zipvoice_zh_en;
{
This file shows how to use the text to speech API of sherpa-onnx
with ZipVoice TTS models.

It generates speech from text and saves it to a wave file.

Please visit
https://k2-fsa.github.io/sherpa/onnx/tts/zipvoice.html
to download the model.
}

{$mode objfpc}

uses
  ctypes,
  SysUtils,
  sherpa_onnx;

function ProgressCallback(Samples: pcfloat; N: cint32; P: cfloat;
  Arg: Pointer): cint32; cdecl;
begin
  WriteLn(Format('Progress: %.2f%%, samples: %d', [P * 100.0, N]));
  Result := 1;
end;

function GetOfflineTts: TSherpaOnnxOfflineTts;
var
  Config: TSherpaOnnxOfflineTtsConfig;
begin
  Config := Default(TSherpaOnnxOfflineTtsConfig);
  Config.Model.ZipVoice.Tokens := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/tokens.txt';
  Config.Model.ZipVoice.Encoder := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/encoder.int8.onnx';
  Config.Model.ZipVoice.Decoder := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/decoder.int8.onnx';
  Config.Model.ZipVoice.Vocoder := './vocos_24khz.onnx';
  Config.Model.ZipVoice.DataDir := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/espeak-ng-data';
  Config.Model.ZipVoice.Lexicon := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/lexicon.txt';
  Config.Model.ZipVoice.FeatScale := 0.1;
  Config.Model.ZipVoice.Tshift := 0.5;
  Config.Model.ZipVoice.TargetRms := 0.1;
  Config.Model.ZipVoice.GuidanceScale := 1.0;
  Config.Model.NumThreads := 2;
  Config.Model.Debug := True;
  Config.MaxNumSentences := 1;

  Result := TSherpaOnnxOfflineTts.Create(Config);
end;

var
  Tts: TSherpaOnnxOfflineTts;
  GenerationConfig: TSherpaOnnxGenerationConfig;
  Wave: TSherpaOnnxWave;
  WaveFilename: AnsiString;
  Audio: TSherpaOnnxGeneratedAudio;
  Text: AnsiString;
  ReferenceText: AnsiString;

begin
  Tts := GetOfflineTts;

  WriteLn('There are ', Tts.GetNumSpeakers, ' speakers');

  Text := '小米的价值观是真诚, 热爱. 真诚，就是不欺人也不自欺. 热爱, 就是全心投入并享受其中.';
  ReferenceText := '那还是三十六年前, 一九八七年. 我呢考上了武汉大学的计算机系.';

  WaveFilename := './sherpa-onnx-zipvoice-distill-int8-zh-en-emilia/test_wavs/leijun-1.wav';
  Wave := SherpaOnnxReadWave(WaveFilename);
  GenerationConfig := Default(TSherpaOnnxGenerationConfig);
  GenerationConfig.SilenceScale := 0.2;
  GenerationConfig.Speed := 1.0;
  GenerationConfig.Sid := 0;
  GenerationConfig.ReferenceAudio := Wave.Samples;
  GenerationConfig.ReferenceAudioLen := Length(Wave.Samples);
  GenerationConfig.ReferenceSampleRate := Wave.SampleRate;
  GenerationConfig.ReferenceText := ReferenceText;
  GenerationConfig.NumSteps := 4;
  GenerationConfig.Extra := '{"min_char_in_sentence": "10"}';

  Audio := Tts.Generate(Text, GenerationConfig, @ProgressCallback, NIL);
  SherpaOnnxWriteWave('./zipvoice-zh-en.wav', Audio.Samples, Audio.SampleRate);
  WriteLn('Saved to ./zipvoice-zh-en.wav');

  FreeAndNil(Tts);
end.

File: dart-api-examples/tts/bin/zipvoice-zh-en.dart
// Copyright (c)  2026  Xiaomi Corporation
import 'dart:io';

import 'package:args/args.dart';
import 'package:sherpa_onnx/sherpa_onnx.dart' as sherpa_onnx;

import './init.dart';

void main(List<String> arguments) async {
  await initSherpaOnnx();

  final parser = ArgParser()
    ..addOption('tokens', help: 'Path to tokens.txt')
    ..addOption('encoder', help: 'Path to the encoder model')
    ..addOption('decoder', help: 'Path to the decoder model')
    ..addOption('vocoder', help: 'Path to the vocoder model')
    ..addOption('data-dir', help: 'Path to espeak-ng-data directory')
    ..addOption('lexicon', help: 'Path to lexicon.txt')
    ..addOption('reference-audio', help: 'Path to reference audio (wav)')
    ..addOption('reference-text', help: 'Reference text for zero-shot TTS')
    ..addOption('text', help: 'Text to generate TTS for')
    ..addOption('output-wav', help: 'Filename to save the generated audio')
    ..addOption(
      'num-steps',
      help: 'Number of inference steps (default: 4)',
      defaultsTo: '4',
    );

  final res = parser.parse(arguments);

  if (res['tokens'] == null ||
      res['encoder'] == null ||
      res['decoder'] == null ||
      res['vocoder'] == null ||
      res['data-dir'] == null ||
      res['lexicon'] == null ||
      res['reference-audio'] == null ||
      res['reference-text'] == null ||
      res['output-wav'] == null ||
      res['text'] == null) {
    print(parser.usage);
    exit(1);
  }

  final tokens = res['tokens'] as String;
  final encoder = res['encoder'] as String;
  final decoder = res['decoder'] as String;
  final vocoder = res['vocoder'] as String;
  final dataDir = res['data-dir'] as String;
  final lexicon = res['lexicon'] as String;
  final referenceAudioPath = res['reference-audio'] as String;
  final referenceText = res['reference-text'] as String;
  final text = res['text'] as String;
  final outputWav = res['output-wav'] as String;
  final numSteps = int.parse(res['num-steps'] as String);

  final zipvoice = sherpa_onnx.OfflineTtsZipVoiceModelConfig(
    tokens: tokens,
    encoder: encoder,
    decoder: decoder,
    vocoder: vocoder,
    dataDir: dataDir,
    lexicon: lexicon,
  );

  final modelConfig = sherpa_onnx.OfflineTtsModelConfig(
    zipvoice: zipvoice,
    numThreads: 2,
    debug: true,
  );

  final config = sherpa_onnx.OfflineTtsConfig(model: modelConfig);

  final tts = sherpa_onnx.OfflineTts(config);

  final wave = sherpa_onnx.readWave(referenceAudioPath);
  if (wave.samples.isEmpty || wave.sampleRate == 0) {
    throw Exception('Failed to read reference audio: $referenceAudioPath');
  }

  final genConfig = sherpa_onnx.OfflineTtsGenerationConfig(
    speed: 1.0,
    referenceAudio: wave.samples,
    referenceSampleRate: wave.sampleRate,
    referenceText: referenceText,
    numSteps: numSteps,
    extra: {'min_char_in_sentence': 10},
  );

  final audio = tts.generateWithConfig(
    text: text,
    config: genConfig,
    onProgress: (samples, progress) {
      print('Progress: ${(progress * 100).toStringAsFixed(2)}%');
      print('Received samples length: ${samples.length}');
      return 1;
    },
  );

  tts.free();

  sherpa_onnx.writeWave(
    filename: outputWav,
    samples: audio.samples,
    sampleRate: audio.sampleRate,
  );

  print('Saved to $outputWav');
}
